tls/s2n_config.c

```diff
@@ -400,22 +368,79 @@ int s2n_config_add_cert_chain_and_key(struct s2n_config *config, const char *cer
         new_node->next = NULL;
         *insert = new_node;
         insert = &new_node->next;
-    } while (s2n_stuffer_data_available(&chain_in_stuffer));
+    } while (s2n_stuffer_data_available(chain_in_stuffer));
 
-    const uint32_t leftover_chain_amount = s2n_stuffer_data_available(&chain_in_stuffer);
-    GUARD(s2n_stuffer_free(&chain_in_stuffer));
     GUARD(s2n_stuffer_free(&cert_out_stuffer));
 
     /* Leftover data at this point means one of two things:
      * A bug in s2n's PEM parsing OR a malformed PEM in the user's chain.
      * Be conservative and fail instead of using a partial chain.
      */
-    if (leftover_chain_amount > 0) {
+    if (s2n_stuffer_data_available(chain_in_stuffer) > 0) {
         S2N_ERROR(S2N_ERR_INVALID_PEM);
     }
-    
     config->cert_and_key_pairs->cert_chain.chain_size = chain_size;
 
+    return 0;
+}
+
+int s2n_config_add_cert_chain(struct s2n_config *config, const char *cert_chain_pem){
+    struct s2n_stuffer chain_in_stuffer;
+
+    /* Turn the chain into a stuffer */
+    GUARD(s2n_stuffer_alloc_ro_from_string(&chain_in_stuffer, cert_chain_pem));
+    int rc = s2n_config_add_cert_chain_from_stuffer(config, &chain_in_stuffer);
+
+    GUARD(s2n_stuffer_free(&chain_in_stuffer));
+
+    return rc;
+
+}
+
+int s2n_config_add_private_key(struct s2n_config *config, const char *private_key_pem){
+    struct s2n_stuffer key_in_stuffer, key_out_stuffer;
+    struct s2n_blob key_blob;
+
+    GUARD(s2n_pkey_zero_init(&config->cert_and_key_pairs->private_key));
+
+    /* Put the private key pem in a stuffer */
+    GUARD(s2n_stuffer_alloc_ro_from_string(&key_in_stuffer, private_key_pem));
+    GUARD(s2n_stuffer_growable_alloc(&key_out_stuffer, strlen(private_key_pem)));
+
+    /* Convert pem to asn1 and asn1 to the private key. Handles both PKCS#1 and PKCS#8 formats */
+    GUARD(s2n_stuffer_private_key_from_pem(&key_in_stuffer, &key_out_stuffer));
+    GUARD(s2n_stuffer_free(&key_in_stuffer));
+    key_blob.size = s2n_stuffer_data_available(&key_out_stuffer);
+    key_blob.data = s2n_stuffer_raw_read(&key_out_stuffer, key_blob.size);
+    notnull_check(key_blob.data);
+
+    /* Get key type and create appropriate key context */
+    GUARD(s2n_asn1der_to_private_key(&config->cert_and_key_pairs->private_key, &key_blob));
+    GUARD(s2n_stuffer_free(&key_out_stuffer));
+
+    return 0;
+
+}
+
+int s2n_config_add_cert_chain_and_key(struct s2n_config *config, const char *cert_chain_pem, const char *private_key_pem)
+{
+    struct s2n_blob mem;
+
+    /* Allocate the memory for the chain and key struct */
+    GUARD(s2n_alloc(&mem, sizeof(struct s2n_cert_chain_and_key)));
+    config->cert_and_key_pairs = (struct s2n_cert_chain_and_key *)(void *)mem.data;
+    config->cert_and_key_pairs->cert_chain.head = NULL;
+
+    config->cert_and_key_pairs->ocsp_status.data = NULL;
```

Comment:
Unrelated to your change, but aren't these NULL/0 assignments for the fields of `ocsp_status` and `sct_list` redundant if we are memset'ing both to 0?

Reply:
Yes, not sure why that's there. Will remove.